Updating service to NodePort type gets port lost

<!-- This form is for bug reports and feature requests ONLY! 

If you're looking for help check [Stack Overflow](https://stackoverflow.com/questions/tagged/kubernetes) and the [troubleshooting guide](https://kubernetes.io/docs/tasks/debug-application-cluster/troubleshooting/).
-->

**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug
/sig network
/cc @thockin @smarterclayton @freehan 
/assign

**What happened**:

1. Create a ClusterIP type service
```
apiVersion: v1
kind: Service
metadata:
  name: foo
spec:
  type: ClusterIP
  ports:
  - port: 53
    name: port-tcp
    protocol: TCP
    targetPort: 6502
  selector:
    bar: baz
```

```
# kubectl describe svc foo 
Name:			foo
Namespace:		default
Labels:			<none>
Annotations:		<none>
Selector:		bar=baz
Type:			ClusterIP
IP:			192.168.100.251
Port:			port-tcp	53/TCP
Endpoints:		<none>
Session Affinity:	None
Events:			<none>
```

2. Edit the service. Change it's type to NodePort and add a Port. 
Note: the new port has same port number but protocol with the existing Port.
```
spec:
  clusterIP: 192.168.100.251
  ports:
  - name: port-tcp
    port: 53
    protocol: TCP
    targetPort: 6502
  ports:
  - name: port-udp
    port: 53
    protocol: UDP
    targetPort: 6502
  selector:
    bar: baz
  sessionAffinity: None
  type: NodePort
```

3. See the updated service. **The TCP Port was eaten.**
```
# kubectl describe svc foo 
Name:			foo
Namespace:		default
Labels:			<none>
Annotations:		<none>
Selector:		bar=baz
Type:			NodePort
IP:			192.168.100.251
Port:			port-udp	53/UDP
NodePort:		port-udp	31494/UDP
Endpoints:		<none>
Session Affinity:	None
Events:			<none>
```

**What you expected to happen**:
The updated service has both TCP and UDP ports.

**How to reproduce it (as minimally and precisely as possible)**:
See the step above.

**Anything else we need to know?**:
If I create a NodePort service with the above two ports, it works.
```
apiVersion: v1
kind: Service
metadata:
  name: foo
spec:
  type: NodePort
  ports:
  - port: 53
    name: port-tcp
    protocol: TCP
    targetPort: 6502
  - port: 53
    name: port-udp
    protocol: UDP
    targetPort: 6502
  selector:
    bar: baz
```

```
# kubectl describe svc foo 
Name:			foo
Namespace:		default
Labels:			<none>
Annotations:		<none>
Selector:		bar=baz
Type:			NodePort
IP:			192.168.100.74
Port:			port-tcp	53/TCP
NodePort:		port-tcp	30680/TCP
Endpoints:		<none>
Port:			port-udp	53/UDP
NodePort:		port-udp	30680/UDP
Endpoints:		<none>
Session Affinity:	None
Events:			<none>
```

I write a test case to verify this case. I got a panic.
```
--- FAIL: TestUpdateNodePort (0.00s)
panic: duplicate node port [recovered]
	panic: duplicate node port

goroutine 116 [running]:
testing.tRunner.func1(0xc42022ca90)
	/usr/local/go/src/testing/testing.go:622 +0x29d
panic(0x143d0c0, 0xc420064c90)
	/usr/local/go/src/runtime/panic.go:489 +0x2cf
k8s.io/kubernetes/pkg/registry/core/service.(*REST).updateNodePort(0xc4202bee10, 0xc4201cde00, 0xc420304f00, 0xc420051e38, 0x0, 0x0)
	/home/cloud/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/core/service/rest.go:616 +0x262
k8s.io/kubernetes/pkg/registry/core/service.TestUpdateNodePort(0xc42022ca90)
	/home/cloud/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/registry/core/service/rest_test.go:1566 +0x1f2b
testing.tRunner(0xc42022ca90, 0x16e9718)
	/usr/local/go/src/testing/testing.go:657 +0x96
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:697 +0x2ca
FAIL	k8s.io/kubernetes/pkg/registry/core/service	0.130s
make: *** [test] Error 1
```

The root cause is as below. _We should take another look at this logic._
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/service/rest.go#L614-L617
```GO
// Detect duplicate node ports; this should have been caught by validation, so we panic
if contains(newNodePorts, nodePort) {
	panic("duplicate node port")
}
```

BTW: #46197 changed some related code but it doesn't change the logic here. So I believe this bug is not introduced by #46197. I also tested this on v1.5.1, it has the same issue.

**Environment**:
- Kubernetes version (use `kubectl version`): HEAD a9bf44101b3ec13a3e94df926c95ba5384d18e28
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

--- a/pkg/registry/core/service/rest.go
+++ b/pkg/registry/core/service/rest.go
@@ -22,7 +22,6 @@ import (
 	"net"
 	"net/http"
 	"net/url"
-	"reflect"
 	"strconv"
 
 	"github.com/golang/glog"
@@ -461,7 +460,7 @@ func containsNumber(haystack []int, needle int) bool {
 // so small that we expect a linear search to be faster
 func containsNodePort(serviceNodePorts []ServiceNodePort, serviceNodePort ServiceNodePort) bool {
 	for _, snp := range serviceNodePorts {
-		if reflect.DeepEqual(snp, serviceNodePort) {
+		if snp == serviceNodePort {
 			return true
 		}
 	}

--- a/pkg/registry/core/service/rest.go
+++ b/pkg/registry/core/service/rest.go
@@ -617,9 +617,8 @@ func (rs *REST) updateNodePorts(oldService, newService *api.Service, nodePortOp 
 			servicePort.NodePort = int32(nodePortNumber)
 			nodePort.NodePort = servicePort.NodePort
 		}
-		// Detect duplicate node ports; this should have been caught by validation, so we panic
 		if containsNodePort(newNodePorts, nodePort) {
-			panic("duplicate node port")
+			return fmt.Errorf("duplicate nodePort: %v", nodePort)
 		}
 		newNodePorts = append(newNodePorts, nodePort)
 	}

--- a/pkg/registry/core/service/rest.go
+++ b/pkg/registry/core/service/rest.go
@@ -22,6 +22,7 @@ import (
 	"net"
 	"net/http"
 	"net/url"
+	"reflect"
 	"strconv"
 
 	"github.com/golang/glog"
@@ -59,6 +60,16 @@ type REST struct {
 	serviceIPs       ipallocator.Interface
 	serviceNodePorts portallocator.Interface
 	proxyTransport   http.RoundTripper
+}
+
+// ServiceNodePort includes protocol and port number of a service NodePort.
+type ServiceNodePort struct {
+	// The IP protocol for this port. Supports "TCP" and "UDP".
+	Protocol api.Protocol
+
+	// The port on each node on which this service is exposed.
+	// Default is to auto-allocate a port if the ServiceType of this Service requires one.
+	NodePort int32
 }
 
 // NewStorage returns a new REST.
@@ -437,9 +448,20 @@ func (rs *REST) ResourceLocation(ctx genericapirequest.Context, id string) (*url
 
 // This is O(N), but we expect haystack to be small;
 // so small that we expect a linear search to be faster
-func contains(haystack []int, needle int) bool {
+func containsNumber(haystack []int, needle int) bool {
 	for _, v := range haystack {
 		if v == needle {
+			return true
+		}
+	}
+	return false
+}
+
+// This is O(N), but we expect serviceNodePorts to be small;
+// so small that we expect a linear search to be faster
+func containsNodePort(serviceNodePorts []ServiceNodePort, serviceNodePort ServiceNodePort) bool {
+	for _, snp := range serviceNodePorts {
+		if reflect.DeepEqual(snp, serviceNodePort) {
 			return true
 		}
 	}
@@ -569,44 +591,49 @@ func (rs *REST) initNodePorts(service *api.Service, nodePortOp *portallocator.Po
 }
 
 func (rs *REST) updateNodePorts(oldService, newService *api.Service, nodePortOp *portallocator.PortAllocationOperation) error {
-	oldNodePorts := CollectServiceNodePorts(oldService)
+	oldNodePortsNumbers := CollectServiceNodePorts(oldService)
+	newNodePorts := []ServiceNodePort{}
+	portAllocated := map[int]bool{}
 
-	newNodePorts := []int{}
 	for i := range newService.Spec.Ports {
 		servicePort := &newService.Spec.Ports[i]
-		nodePort := int(servicePort.NodePort)
-		if nodePort != 0 {
-			if !contains(oldNodePorts, nodePort) {
-				err := nodePortOp.Allocate(nodePort)
+		nodePort := ServiceNodePort{Protocol: servicePort.Protocol, NodePort: servicePort.NodePort}
+		if nodePort.NodePort != 0 {
+			if !containsNumber(oldNodePortsNumbers, int(nodePort.NodePort)) && !portAllocated[int(nodePort.NodePort)] {
+				err := nodePortOp.Allocate(int(nodePort.NodePort))
 				if err != nil {
-					el := field.ErrorList{field.Invalid(field.NewPath("spec", "ports").Index(i).Child("nodePort"), nodePort, err.Error())}
+					el := field.ErrorList{field.Invalid(field.NewPath("spec", "ports").Index(i).Child("nodePort"), nodePort.NodePort, err.Error())}
 					return errors.NewInvalid(api.Kind("Service"), newService.Name, el)
 				}
+				portAllocated[int(nodePort.NodePort)] = true
 			}
 		} else {
-			nodePort, err := nodePortOp.AllocateNext()
+			nodePortNumber, err := nodePortOp.AllocateNext()
 			if err != nil {
 				// TODO: what error should be returned here?  It's not a
 				// field-level validation failure (the field is valid), and it's
 				// not really an internal error.
 				return errors.NewInternalError(fmt.Errorf("failed to allocate a nodePort: %v", err))
 			}
-			servicePort.NodePort = int32(nodePort)
+			servicePort.NodePort = int32(nodePortNumber)
+			nodePort.NodePort = servicePort.NodePort
 		}
 		// Detect duplicate node ports; this should have been caught by validation, so we panic
-		if contains(newNodePorts, nodePort) {
+		if containsNodePort(newNodePorts, nodePort) {
 			panic("duplicate node port")
 		}
 		newNodePorts = append(newNodePorts, nodePort)
 	}
 
+	newNodePortsNumbers := CollectServiceNodePorts(newService)
+
 	// The comparison loops are O(N^2), but we don't expect N to be huge
 	// (there's a hard-limit at 2^16, because they're ports; and even 4 ports would be a lot)
-	for _, oldNodePort := range oldNodePorts {
-		if contains(newNodePorts, oldNodePort) {
+	for _, oldNodePortNumber := range oldNodePortsNumbers {
+		if containsNumber(newNodePortsNumbers, oldNodePortNumber) {
 			continue
 		}
-		nodePortOp.ReleaseDeferred(oldNodePort)
+		nodePortOp.ReleaseDeferred(int(oldNodePortNumber))
 	}
 
 	return nil

--- a/pkg/registry/core/service/rest_test.go
+++ b/pkg/registry/core/service/rest_test.go
@@ -1570,6 +1570,92 @@ func TestUpdateNodePorts(t *testing.T) {
 			},
 			expectSpecifiedNodePorts: []int{},
 		},
+		{
+			name: "Add new ServicePort with a different protocol without changing port numbers",
+			oldService: &api.Service{
+				ObjectMeta: metav1.ObjectMeta{Name: "foo"},
+				Spec: api.ServiceSpec{
+					Selector:        map[string]string{"bar": "baz"},
+					SessionAffinity: api.ServiceAffinityNone,
+					Type:            api.ServiceTypeNodePort,
+					Ports: []api.ServicePort{
+						{
+							Name:       "port-tcp",
+							Port:       53,
+							TargetPort: intstr.FromInt(6502),
+							Protocol:   api.ProtocolTCP,
+							NodePort:   30053,
+						},
+					},
+				},
+			},
+			newService: &api.Service{
+				ObjectMeta: metav1.ObjectMeta{Name: "foo"},
+				Spec: api.ServiceSpec{
+					Selector:        map[string]string{"bar": "baz"},
+					SessionAffinity: api.ServiceAffinityNone,
+					Type:            api.ServiceTypeNodePort,
+					Ports: []api.ServicePort{
+						{
+							Name:       "port-tcp",
+							Port:       53,
+							TargetPort: intstr.FromInt(6502),
+							Protocol:   api.ProtocolTCP,
+							NodePort:   30053,
+						},
+						{
+							Name:       "port-udp",
+							Port:       53,
+							TargetPort: intstr.FromInt(6502),
+							Protocol:   api.ProtocolUDP,
+							NodePort:   30053,
+						},
+					},
+				},
+			},
+			expectSpecifiedNodePorts: []int{30053, 30053},
+		},
+		{
+			name: "Change service type from ClusterIP to NodePort with same NodePort number but different protocols",
+			oldService: &api.Service{
+				ObjectMeta: metav1.ObjectMeta{Name: "foo"},
+				Spec: api.ServiceSpec{
+					Selector:        map[string]string{"bar": "baz"},
+					SessionAffinity: api.ServiceAffinityNone,
+					Type:            api.ServiceTypeClusterIP,
+					Ports: []api.ServicePort{{
+						Port:       53,
+						Protocol:   api.ProtocolTCP,
+						TargetPort: intstr.FromInt(6502),
+					}},
+				},
+			},
+			newService: &api.Service{
+				ObjectMeta: metav1.ObjectMeta{Name: "foo"},
+				Spec: api.ServiceSpec{
+					Selector:        map[string]string{"bar": "baz"},
+					SessionAffinity: api.ServiceAffinityNone,
+					Type:            api.ServiceTypeNodePort,
+					Ports: []api.ServicePort{
+						{
+							Name:       "port-tcp",
+							Port:       53,
+							TargetPort: intstr.FromInt(6502),
+							Protocol:   api.ProtocolTCP,
+							NodePort:   30053,
+						},
+						{
+							Name:       "port-udp",
+							Port:       53,
+							TargetPort: intstr.FromInt(6502),
+							Protocol:   api.ProtocolUDP,
+							NodePort:   30053,
+						},
+					},
+				},
+			},
+			expectSpecifiedNodePorts: []int{30053, 30053},
+		},
 	}
 
 	for _, test := range testCases {

--- a/test/e2e/network/service.go
+++ b/test/e2e/network/service.go
@@ -803,6 +803,53 @@ var _ = SIGDescribe("Services", func() {
 		}
 	})
 
+	It("should be able to update NodePorts with two same port numbers but different protocols", func() {
+		serviceName := "nodeport-update-service"
+		ns := f.Namespace.Name
+		jig := framework.NewServiceTestJig(cs, serviceName)
+
+		By("creating a TCP service " + serviceName + " with type=ClusterIP in namespace " + ns)
+		tcpService := jig.CreateTCPServiceOrFail(ns, nil)
+		defer func() {
+			framework.Logf("Cleaning up the updating NodePorts test service")
+			err := cs.Core().Services(ns).Delete(serviceName, nil)
+			Expect(err).NotTo(HaveOccurred())
+		}()
+		jig.SanityCheckService(tcpService, v1.ServiceTypeClusterIP)
+		svcPort := int(tcpService.Spec.Ports[0].Port)
+		framework.Logf("service port TCP: %d", svcPort)
+
+		// Change the services to NodePort and add a UDP port.
+
+		By("changing the TCP service to type=NodePort and add a UDP port")
+		newService := jig.UpdateServiceOrFail(ns, tcpService.Name, func(s *v1.Service) {
+			s.Spec.Type = v1.ServiceTypeNodePort
+			s.Spec.Ports = []v1.ServicePort{
+				{
+					Name:     "tcp-port",
+					Port:     80,
+					Protocol: v1.ProtocolTCP,
+				},
+				{
+					Name:     "udp-port",
+					Port:     80,
+					Protocol: v1.ProtocolUDP,
+				},
+			}
+		})
+		jig.SanityCheckService(newService, v1.ServiceTypeNodePort)
+		if len(newService.Spec.Ports) != 2 {
+			framework.Failf("new service should have two Ports")
+		}
+		for _, port := range newService.Spec.Ports {
+			if port.NodePort == 0 {
+				framework.Failf("new service failed to allocate NodePort for Port %s", port.Name)
+			}
+
+			framework.Logf("new service allocates NodePort %d for Port %s", port.NodePort, port.Name)
+		}
+	})
+
 	It("should be able to change the type from ExternalName to ClusterIP", func() {
 		serviceName := "externalname-service"
 		ns := f.Namespace.Name

